scripts/src/services/orders.ts

```diff
@@ -80,54 +96,70 @@ export async function submitEarn(
+		}
 	}
 
-	const offer: offerDb.Offer = await offerDb.Offer.findOneById(openOrder.offerId);
-	// create a transaction Order
-	const order = Object.assign(new db.Order(), {
-		id: openOrder.id,
-		userId: openOrder.userId,
-		offerId: openOrder.offerId,
-		amount: offer.amount,
-		type: "earn",
-		status: "pending",
-		meta: offer.meta.order_meta,
-	});
+	// transition open order to pending order
+	const order = new db.Order(openOrder, offer);
 	offer.cap.used += 1;
 	await offer.save();
 	await order.save();
 	openOrdersDB.delete(openOrder.id);
 
-	await payment.payTo(walletAddress, appId, order.amount, order.id, logger);
+	// pay or start timer for payment
+	if (offer.type === "earn") {
+		await payment.payTo(walletAddress, appId, offer.amount, order.id, logger);
+	} else {
+		await submitSpend(order, offer, walletAddress, appId, logger);
+	}
 
 	return orderDbToApi(order, logger);
 }
 
-function orderDbToApi(order: db.Order, logger: LoggerInstance): Order {
-	return {
-		status: order.status,
-		result: order.value,
+export async function submitSpend(
+	order: db.Order, offer: offerDb.Offer, walletAddress: string, appId: string, logger: LoggerInstance): Promise<void> {
+	async function makeFailed(orderId: string) {
+		// XXX lock on order.id
+		const order = await db.Order.findOneById(orderId);
+		order.status = "failed";
+		const offer = await offerDb.Offer.findOneById(order.offerId);
+		offer.cap.used -= 1;
+		await offer.save();
+		await order.save();
+	}
+
+	// start a timer for order.expiration + grace till this order becomes failed
+	// setTimeout(makeFailed, order.expiration, order.id);
+
+	// simulate payment complete // XXX delete this
+	const payment: CompletedPayment = {
 		id: order.id,
-		completion_date: (order.completionDate || order.createdDate).toISOString(), // XXX should we separate the dates?
-		blockchain_data: order.blockchainData,
-		offer_type: order.type,
-		title: order.meta.title,
-		description: order.meta.description,
-		call_to_action: order.meta.call_to_action,
-		content: order.meta.content,
-		amount: order.amount,
+		app_id: appId,
+		transaction_id: "some transaction",
+		recipient_address: offer.blockchainData.recipient_address, // offer received the kin
+		sender_address: walletAddress, // user sent the kin
+		amount: offer.amount,
+		timestamp: (new Date()).toISOString(),
 	};
-}
 
-export async function submitSpend(orderId: string, logger: LoggerInstance): Promise<void> {
+	paymentComplete(payment, logger);
 	return;
 }
 
-export async function cancelOrder(options, logger: LoggerInstance): Promise<void> {
-	return;
+export async function cancelOrder(orderId: string, logger: LoggerInstance): Promise<void> {
+	// you can only delete an open order - not a pending order
+	// validate order
+	const openOrder: db.OpenOrder = openOrdersDB.get(orderId);
+	if (!openOrder) {
+		throw Error(`no such order ${orderId}`);
+	}
+	openOrdersDB.delete(openOrder.id);
```

Comment:
It'a minute but it's a bit better (best practice) to use `orderId`, they are obviously the same at this point and the compiler won't have to look up the property on the object.

Reply:
is this just an optimization or a javascript best practice?